### PR TITLE
Search latest

### PR DESCRIPTION
**Pull request by whytheplatypus**
Joining notices and documents allows queries to deal only with the latest version of a document.

relates to: https://github.com/CMSgov/cmcs-eregulations/pull/62

### CODE REVIEW

File: regcore/db/django_models.py [hunk 1]
```diff
@@ -119,10 +121,12 @@ def listing(self, doc_type, label=None):
...
 
-        query = query.only('version', 'label_string').order_by('version')
-        # Flattens
-        versions = [v for v in query.values_list('version', 'label_string')]
-        return versions
+        query = query.values_list('version', 'label_string')
+        notice = Notice.objects.filter(document_number=OuterRef('version'))
+        query = query.annotate(effective_on=Subquery(notice.values('effective_on')[:1]))
+        query = query.order_by('label_string', '-effective_on')
+        query = query.distinct('label_string')
```
Comment (whytheplatypus) [inline regcore/db/django_models.py]: This was a POC, should probably be pulled out for now

File: regcore/db/django_models.py [hunk 2]
```diff
@@ -119,10 +121,12 @@ def listing(self, doc_type, label=None):
             query = Document.objects.filter(
                 doc_type=doc_type, label_string=label)
 
-        query = query.only('version', 'label_string').order_by('version')
-        # Flattens
-        versions = [v for v in query.values_list('version', 'label_string')]
-        return versions
+        query = query.values_list('version', 'label_string')
+        notice = Notice.objects.filter(document_number=OuterRef('version'))
```
Comment (whytheplatypus) [inline regcore/db/django_models.py]: This needs to be ordered with `-effective_on`

File: regcore_pgsql/views.py [hunk 3]
```diff
@@ -20,6 +22,11 @@ def matching_sections(search_args):
     if search_args.regulation:
         sections_query = sections_query.filter(
             documentindex__doc_root=search_args.regulation)
+
+    notice = Notice.objects.filter(document_number=OuterRef('version'))
```
Comment (whytheplatypus) [inline regcore_pgsql/views.py]: This needs to be ordered with `-effective_on`

File: regcore_pgsql/views.py [hunk 4]
```diff
@@ -20,6 +22,11 @@ def matching_sections(search_args):
...
+    sections_query = sections_query.annotate(effective_on=Subquery(notice.values('effective_on')[:1]))
+    sections_query = sections_query.order_by('label_string', '-effective_on', '-rank')
+    sections_query = sections_query.distinct('label_string')
```
Comment (whytheplatypus) [inline regcore_pgsql/views.py]: Might be able to make this faster by doing this distinct filter first?

File: regcore/db/django_models.py [hunk 5]
```diff
@@ -119,10 +121,12 @@ def listing(self, doc_type, label=None):
             query = Document.objects.filter(
                 doc_type=doc_type, label_string=label)
 
-        query = query.only('version', 'label_string').order_by('version')
-        # Flattens
-        versions = [v for v in query.values_list('version', 'label_string')]
-        return versions
+        query = query.values_list('version', 'label_string')
```
Comment (theabrad) [inline regcore/db/django_models.py:124]: pull this out into it's own function to describe what it is doing

File: regcore_pgsql/views.py (hunk 3 above)
Comment (theabrad) [reply regcore_pgsql/views.py]: make this into it's own function

**Review by theabrad — APPROVED** [review]

**Review by cgodwin1 — APPROVED** [review]

**Pull request merged**